### PR TITLE
[Serve] Http proxy & router & handle to support multiplex impl

### PR DESCRIPTION
**Pull request by sihanwang41**
## Why are these changes needed?

- Support handle.options(multiplexed_model_id="")
- Http proxy to extract model id on the fly
- Choose correct replica based on information.
- nit: Move handle metrics pusher to router.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/_private/router.py [hunk 1]
```diff
@@ -305,6 +348,32 @@ def __init__(
             call_in_event_loop=event_loop,
         )
 
+        self.deployment_name = deployment_name
```
Comment (sihanwang41) [inline python/ray/serve/_private/router.py:351]: no change in functionality. move logic to router

**Comment by sihanwang41** [comment]
No tests added. wait for the previous [pr](https://github.com/ray-project/ray/pull/35372) merged.

File: python/ray/serve/_private/autoscaling_metrics.py [hunk 2]
```diff
@@ -1,88 +1,14 @@
...
-from typing import Callable, DefaultDict, Dict, List, Optional, Type
+from typing import DefaultDict, Dict, List, Optional
 
-import ray
 from ray.serve._private.constants import SERVE_LOGGER_NAME
 
 logger = logging.getLogger(SERVE_LOGGER_NAME)
 
 
-def start_metrics_pusher(
```
Comment (sihanwang41) [inline python/ray/serve/_private/autoscaling_metrics.py:16]: move it to the utils.py

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -137,51 +159,100 @@ def update_running_replicas(self, running_replicas: List[RunningReplicaInfo]):
             self._reset_replica_iterator()
             self.config_updated_event.set()
 
+    def _assign_replica(self, query: Query, replica: RunningReplicaInfo):
```
Comment (sihanwang41) [inline python/ray/serve/_private/router.py:162]: No change, move it out of the `_try_assign_replica`

File: python/ray/serve/_private/router.py (hunk 1 above)
Comment (sihanwang41) [reply python/ray/serve/_private/router.py:380]: 1. Move Metrics pusher logic lifecycle out of the Router class.
2. Router will pass necessary arguments to initiate metrics pusher and start the pusher.

File: python/ray/serve/_private/utils.py [hunk 4]
```diff
@@ -569,3 +577,77 @@ def extract_self_if_method_call(args: List[Any], func: Callable) -> Optional[obj
                 return args[0]
 
     return None
+
+
+class MetricsPusher:
```
Comment (sihanwang41) [inline python/ray/serve/_private/utils.py:582]: Barely no change, and move it outside the autoscaling_policy.py. and start the MetricsPusher to manage lifecycle with clear interface.

**Review by edoakes — APPROVED** [review]
LGTM.

We really need to add some tests for the scheduling logic in `router.py`, but not going to block this PR on it. We can work together on that as I need to make some changes for streaming too.

**Pull request merged**